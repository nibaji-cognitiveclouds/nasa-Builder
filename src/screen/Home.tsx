/** @format */

import { useNavigation } from "@react-navigation/native";
import { StackNavigationProp } from "@react-navigation/stack";
import axios from "axios";
import React from "react";
import { ActivityIndicator, Button, View } from "react-native";
import { TextInput } from "react-native-gesture-handler";

export type RootStackParamList = {
	Details: {};
};

const Home: React.FC = () => {
	const [id, setId] = React.useState("");
	const [loading, setLoading] = React.useState(false);
	const navigation = useNavigation<StackNavigationProp<RootStackParamList>>();

	function getRandom() {
		axios
			.get(
				"https://api.nasa.gov/neo/rest/v1/neo/browse?api_key=cgFoDzaE9hdKE9UousiEPaX0V0B8GaO0ICUOMexs"
			)
			.then((res) => {
				const random = Math.floor(
					Math.random() * res.data?.near_earth_objects?.length
				);
				navigation.navigate("Details", {
					id: res.data?.near_earth_objects?.[random]?.id,
				});
			})
			.finally(() => {
				setLoading(false);
			});
	}

	return (
		<View style={{ justifyContent: "center", alignItems: "center" }}>
			<TextInput
				placeholder="Enter asteroid id"
				placeholderTextColor={"grey"}
				onChangeText={(text) => setId(text)}
				style={{
					padding: 5,
					borderColor: "grey",
					borderWidth: 1,
					borderRadius: 5,
					margin: 5,
					width: "70%",
					marginBottom: 5,
				}}
			/>
			<View
				style={{
					flexDirection: "row",
					justifyContent: "space-around",
					alignItems: "center",
				}}
			>
				<Button
					disabled={!id.length}
					title="Submit"
					onPress={() => navigation.navigate("Details", { id: id })}
				/>
				{loading ? (
					<ActivityIndicator />
				) : (
					<Button
						title="Random Asteroid"
						onPress={() => {
							setLoading(true);
							getRandom();
						}}
					/>
				)}
			</View>
		</View>
	);
};

export default Home;
