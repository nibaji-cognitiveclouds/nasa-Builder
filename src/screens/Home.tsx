/** @format */

import { useNavigation } from "@react-navigation/native";
import { NativeStackNavigationProp } from "@react-navigation/native-stack";
import axios from "axios";
import React, { FC, useState } from "react";
import {
	ActivityIndicator,
	Alert,
	Button,
	SafeAreaView,
	TextInput,
	View,
} from "react-native";
import { styles } from "../styles/screens";
import { routes } from "../types/navigation";

const Home: FC = () => {
	const [text, setText] = useState<string>("");
	const [loading, setLoading] = useState<boolean>(false);

	const navigation = useNavigation<NativeStackNavigationProp<routes>>();

	const getRandomAsteroid = () => {
		setLoading(true);
		axios
			.get("https://api.nasa.gov/neo/rest/v1/neo/browse?api_key=DEMO_KEY")
			.then((res) => {
				const randomNumber = Math.floor(
					Math.random() * res.data.near_earth_objects.length
				);
				navigation.navigate("Details", {
					id: res.data.near_earth_objects[randomNumber].id,
				});
			})
			.catch(() => {
				Alert.alert("Error", "Something Went wrong", [
					{
						text: "ok",
					},
				]);
			})
			.finally(() => setLoading(false));
	};

	return (
		<SafeAreaView testID="home" style={styles.container}>
			<TextInput
				placeholder="Enter Asteroid ID"
				placeholderTextColor={"grey"}
				onChangeText={(text) => setText(text)}
				style={styles.input}
			/>
			<View style={styles.buttonsRow}>
				<Button
					disabled={text.length == 0}
					title="Submit"
					onPress={() => {
						navigation.navigate("Details", { id: text });
					}}
				/>
				{loading ? (
					<ActivityIndicator color={"red"} size={20} />
				) : (
					<Button
						disabled={loading}
						title="Random Asteroid"
						onPress={() => {
							getRandomAsteroid();
						}}
					/>
				)}
			</View>
		</SafeAreaView>
	);
};

export default Home;
