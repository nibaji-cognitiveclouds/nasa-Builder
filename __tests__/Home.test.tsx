/** @format */

import { render } from "@testing-library/react-native";
import React from "react";
import Home from "../src/screens/Home";

jest.mock("@react-navigation/native", () => {
	return {
		useNavigation: jest.fn(),
	};
});

describe("Home", () => {
	it("renders without crash", () => {
		render(<Home />);
	});

	it("matches snapshot", () => {
		const tree = render(<Home />);
		expect(tree.toJSON()).toMatchSnapshot();
	});

	it("has component with req testID", () => {
		const tree = render(<Home />);
		expect(tree.getByTestId("home")).toBeTruthy();
	});

	it("has button with req title", () => {
		const tree = render(<Home />);
		expect(tree.getByText("Random Asteroid")).toBeTruthy();
	});

	it("has textinput with req placeholder", () => {
		const tree = render(<Home />);
		expect(tree.getByPlaceholderText("Enter Asteroid ID")).toBeTruthy();
	});
});
