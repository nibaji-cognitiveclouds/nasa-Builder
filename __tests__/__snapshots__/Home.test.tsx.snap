// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Home matches snapshot 1`] = `
<RCTSafeAreaView
  emulateUnlessSupported={true}
  style={
    Object {
      "alignItems": "center",
      "backgroundColor": "white",
      "flex": 1,
      "justifyContent": "center",
      "padding": 10,
    }
  }
  testID="home"
>
  <TextInput
    allowFontScaling={true}
    onChangeText={[Function]}
    placeholder="Enter Asteroid ID"
    placeholderTextColor="grey"
    rejectResponderTermination={true}
    style={
      Object {
        "borderColor": "grey",
        "borderRadius": 10,
        "borderWidth": 1,
        "marginBottom": 10,
        "padding": 10,
        "width": "80%",
      }
    }
    underlineColorAndroid="transparent"
  />
  <View
    style={
      Object {
        "alignItems": "center",
        "flexDirection": "row",
        "justifyContent": "space-between",
      }
    }
  >
    <View
      accessibilityRole="button"
      accessibilityState={
        Object {
          "disabled": true,
        }
      }
      accessible={true}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "opacity": 1,
        }
      }
    >
      <View
        style={
          Array [
            Object {},
            Object {},
          ]
        }
      >
        <Text
          disabled={true}
          style={
            Array [
              Object {
                "color": "#007AFF",
                "fontSize": 18,
                "margin": 8,
                "textAlign": "center",
              },
              Object {
                "color": "#cdcdcd",
              },
            ]
          }
        >
          Submit
        </Text>
      </View>
    </View>
    <View
      accessibilityRole="button"
      accessibilityState={Object {}}
      accessible={true}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "opacity": 1,
        }
      }
    >
      <View
        style={
          Array [
            Object {},
          ]
        }
      >
        <Text
          disabled={false}
          style={
            Array [
              Object {
                "color": "#007AFF",
                "fontSize": 18,
                "margin": 8,
                "textAlign": "center",
              },
            ]
          }
        >
          Random Asteroid
        </Text>
      </View>
    </View>
  </View>
</RCTSafeAreaView>
`;
